### PR TITLE
CF 1.8 geometries

### PR DESCRIPTION
**Pull request by benjwadams**
Draft implementation of CF 1.8 geometries.  @greenteadigital, please review both this and the relevant spec section to see if it's up to snuff.  Please don't merge until we decide all's OK.

### CODE REVIEW

File: compliance_checker/cf/cf.py [hunk 1]
```diff
@@ -119,7 +119,6 @@ def __init__(self, options=None):
             "7.2": "§7.2 Cell Measures",
             "7.3": "§7.3 Cell Methods",
             "7.4": "§7.4 Climatological Statistics",
-            "7.5": "§7.5 Geometries",
```
Comment (greenteadigital) [inline compliance_checker/cf/cf.py:122]: I'd like to see this go back in. It's documentary, and having a 7.5 implementation in the codebase but not adding it to this map seems misleading.

Comment (benjwadams) [reply compliance_checker/cf/cf.py:122]: The various new appendix entries follow the pattern of extending the respective CF checker classes.   This change follows the convention previously established by the checker classes and also makes clear when a new section has been introduced.  CF < 1.8 does not have a check for section 7.5 and this makes that clear.

Comment (greenteadigital) [reply compliance_checker/cf/cf.py:122]: OK, gotcha.

Comment (benjwadams) [reply compliance_checker/cf/cf.py:122]: Also, if a section is ever renamed or moved in a release of CF (to my knowledge this hasn't occurred yet), we will have to put the updated dict entry in the new class anyhow.

File: compliance_checker/cf/cf_18.py [hunk 2]
```diff
@@ -108,6 +114,259 @@ def check_taxa(self, ds: MemoizedDataset):
...
+            # If any variables weren't found, we can't continue
+            if not_found_node_vars:
+                geom_valid.messages.append(
+                                  "The following referenced node coordinate"
+                                  "variables for geometry variable"
+                                  f'"{geometry_var_name}" were not found: '
+                                  f'{not_found_node_vars}')
+                results.append(geom_valid.to_result())
+                continue
+                return error_msgs
```
Comment (greenteadigital) [inline compliance_checker/cf/cf_18.py:173]: the return statement is unreachable code, pls remove or clarify intent

File: compliance_checker/cf/cf_18.py [hunk 3]
```diff
@@ -2,6 +2,12 @@
 from compliance_checker import MemoizedDataset
 from compliance_checker.cf.cf import CF1_7Check
 from netCDF4 import Dataset
+from compliance_checker.base import BaseCheck, BaseNCCheck, Result, TestCtx
+from shapely.geometry import (MultiPoint, LineString, MultiLineString, Polygon,
+                              MultiPolygon)
```
Comment (greenteadigital) [inline compliance_checker/cf/cf_18.py:7]: only `Polygon` is used

File: compliance_checker/cf/cf_18.py [hunk 4]
```diff
@@ -108,6 +114,259 @@ def check_taxa(self, ds: MemoizedDataset):
...
+class PointGeometry(GeometryStorage):
+    """Class for validating Point/MultiPoint geometries"""
+
+    def check_geometry(self):
+        super().check_geometry()
+        # non-multipoint should have exactly one feature
+        if self.node_count is None:
+            expected_node_count = 1
+        else:
+            expected_node_count = self.node_count
```
Comment (greenteadigital) [inline compliance_checker/cf/cf_18.py:243]: `expected_node_count` is never used, remove?

File: compliance_checker/cf/cf_18.py [hunk 5]
```diff
@@ -108,6 +114,259 @@ def check_taxa(self, ds: MemoizedDataset):
...
+        if messages:
+            return messages
+        if self.part_node_count is not None:
+            extents = np.concatenate([np.array([0]),
+                                     self.part_node_count[:].cumsum()])
+            if self.interior_ring is not None:
+                ring_orientation = self.interior_ring[:].astype(bool)
+            else:
+                ring_orientation = np.zeros(len(self.part_count), dtype=bool)
+            current_node_count = self.node_count[:].copy()
```
Comment (greenteadigital) [inline compliance_checker/cf/cf_18.py:342]: `current_node_count` is never used

File: compliance_checker/cf/cf_18.py [hunk 6]
```diff
@@ -108,6 +114,259 @@ def check_taxa(self, ds: MemoizedDataset):
...
+            node_indexer_len = len(self.part_node_count)
+        else:
+            extents = np.concatenate([np.array([0]),
+                                     self.node_count[:].cumsum()])
+            node_indexer_len = len(self.node_count)
+            ring_orientation = np.zeros(node_indexer, dtype=bool)
```
Comment (greenteadigital) [inline compliance_checker/cf/cf_18.py:348]: `node_indexer` is undefined

File: compliance_checker/cf/cf_18.py [hunk 7]
```diff
@@ -108,6 +114,259 @@ def check_taxa(self, ds: MemoizedDataset):
...
+            if geometry_var_name not in ds.variables:
+                geom_valid.messages.append("Cannot find geometry variable "
+                                           f"named {geometry_var_name}")
+                results.append(geom_valid.to_result())
+                continue
+            else:
+                geometry_var = ds.variables[geometry_var_name]
+
+            geometry_type = getattr(geometry_var, "geometry_type")
+            valid_geometry_types = {"point", "line", "polygon"}
```
Comment (greenteadigital) [inline compliance_checker/cf/cf_18.py:142]: `valid_geometry_types` is never used

File: compliance_checker/tests/test_cf.py [hunk 8]
```diff
@@ -2246,6 +2237,64 @@ def test_check_add_offset_scale_factor_type(self):
...
+        geom_var = dataset.createVariable("geometry", "i4", ())
+        geom_var.geometry_type = "point"
+        geom_var.node_coordinates = "x y"
+        x[:] = 1
+        y[:] = 1
+        self.cf.check_geometry(dataset)
+
+    def test_point_geometry_multiple(self):
+        dataset = MockTimeSeries()
+        point_count = dataset.createDimension("point_count", 3)
```
Comment (greenteadigital) [inline compliance_checker/tests/test_cf.py:2260]: `point_count` is never used

File: compliance_checker/tests/test_cf.py [hunk 9]
```diff
@@ -2246,6 +2237,64 @@ def test_check_add_offset_scale_factor_type(self):
...
+        x = dataset.createVariable("x", "f8", ("point_count",))
+        y = dataset.createVariable("y", "f8", ("point_count",))
+        geom_var = dataset.createVariable("geometry", "i4", ())
+        geom_var.geometry_type = "point"
+        geom_var.node_coordinates = "x y"
+        x[:] = np.array([10, 20, 30])
+        y[:] = np.array([30, 35, 21])
+        results = self.cf.check_geometry(dataset)
+        assert results[0].value[0] == results[0].value[1]
+        point_count_2 = dataset.createDimension("point_count_2", 2)
```
Comment (greenteadigital) [inline compliance_checker/tests/test_cf.py:2272]: `point_count_2` is never used

**Comment by benjwadams** [comment]
@greenteadigital, thanks for the review thus far.  My linter should have caught a lot of those but for some reason was not triggered.

File: .github/workflows/cc-checker-ugrid-test.yml [hunk 10]
```diff
@@ -22,20 +22,20 @@ jobs:
       shell: bash -l {0}
       run: |
         conda create --name TEST python=${{ matrix.python-version }} python=3 pip --file requirements.txt --file test_requirements.txt --strict-channel-priority
-        source activate TEST
+        conda activate TEST
```
[the remark this answers is not in the document]
Comment (ocefpaf) [reply .github/workflows/cc-checker-ugrid-test.yml]: That is for users but for CIs a more "crude" source is more robust b/c we can skip the `conda init` step.

File: .github/workflows/default-tests.yml [hunk 11]
```diff
@@ -28,7 +28,7 @@ jobs:
       run: |
         conda create --name TEST python=${{ matrix.python-version }} pip "libnetcdf<4.8.0" --file requirements.txt --file test_requirements.txt --strict-channel-priority
         source activate TEST
-        pip install -e . --no-deps --force-reinstall
+        pip install --user -e . --no-deps --force-reinstall
```
[the remark this answers is not in the document]
Comment (ocefpaf) [reply .github/workflows/default-tests.yml]: Got it. But the original problem was a conflict, not a permission, right? Then the `conda activate` brought the permission one b/c the env was not activated. My last commits seems to fix it now we need to discuss if that is the way we want to go b/c regenerating the ncfiles can be costly.

**Pull request merged**